data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultCosmosRepositoryOperations.java

```diff
@@ -253,12 +254,58 @@ public Optional<Number> executeUpdate(PreparedQuery<?, Number> preparedQuery) {
 
     @Override
     public <T> int delete(DeleteOperation<T> operation) {
+        RuntimePersistentEntity<T> persistentEntity = runtimeEntityRegistry.getEntity(operation.getRootEntity());
+        T entity = operation.getEntity();
+        CosmosContainer container = getContainer(persistentEntity);
+        ObjectNode tree = serializeToTree(entity, Argument.of(operation.getRootEntity()));
+        CosmosItemRequestOptions options = new CosmosItemRequestOptions();
+        CosmosItemResponse cosmosItemResponse = container.deleteItem(tree, options);
+        if (cosmosItemResponse.getStatusCode() == HttpResponseStatus.NO_CONTENT.code()) {
+            return 1;
+        }
         return 0;
     }
 
     @Override
     public <T> Optional<Number> deleteAll(DeleteBatchOperation<T> operation) {
-        return Optional.empty();
+        RuntimePersistentEntity<T> persistentEntity = runtimeEntityRegistry.getEntity(operation.getRootEntity());
+        CosmosContainer container = getContainer(persistentEntity);
+        CosmosItemRequestOptions options = new CosmosItemRequestOptions();
+        Argument<T> arg = Argument.of(operation.getRootEntity());
+        Iterator<T> iterator = operation.iterator();
+        int deletedCount = 0;
+        while (iterator.hasNext()) {
```

Comment:
It should be possible to create X batching operation and run them together

Reply:
I didn't find API to do it here

https://azuresdkdocs.blob.core.windows.net/$web/java/azure-cosmos/latest/com/azure/cosmos/CosmosContainer.html

Do you  know which method would that be?

Reply:
Ok, found executeBulkOperations method and I will change code to use it. Thanks

Reply:
Done here https://github.com/micronaut-projects/micronaut-data/pull/1784/commits/0c2250e0222efe5b4880c418d22887dab2c94a75

However, executeBulkOperations requires PartitionKey so we still have to delete one by one document if PartitionKey not provided.

Reply:
I think we can use PartitionKey.NONE https://azuresdkdocs.blob.core.windows.net/$web/java/azure-cosmos/latest/com/azure/cosmos/models/PartitionKey.html

Reply:
I tried that already, but get an error 
```
PartitionKey has fewer components than defined the collection resource.
```
because partition key is defined on the collection and we are passing None so this would not work.

Reply:
Well if it’s defined then pass it.

Reply:
I mean defined on container level during creation and here we don't get it as an argument. In that case we cannot use it, deletion won't work and if we pass PartitionKey.None it throws error above.

Reply:
When we want to delete multiple or a single instance of the entity, we need to pass the partition key that is defined for that entity.

Reply:
That would be nice, but in order to do that we would have to implement new DeleteInterceptor or DeleteAllInterceptor for Cosmos Db, right? I raised this issue multiple times, our operations classes don't support this and only way to pass partition key is via type roles in prepared query. But in delete(entity) and deleteAll() there is no prepared query.

Reply:
You need to access `PersistentEntity` and get the partition key defined there, if specified by the annotation, or from the repository class etc

Reply:
Ok, I have updated code. If PartitionKey is provided (via type role/prepared query) then we will use it and otherwise will try to calculate form the item being deleted, if partition key defined in config. If partition key not defined (should not happen), then fallback and delete items one by one.

Reply:
It sounds strange that we need to do the one-by-one delete. I would just let it fail. Users might want to expect batch to be used for the batch operation, but I'm fine with it if you want to keep it this way.

Reply:
I would leave it now and just add log statement.